libs/core/async_combinators/include/hpx/async_combinators/when_any.hpp

```diff
@@ -367,112 +373,124 @@ namespace hpx { namespace lcos {
+        lcos::local::futures_factory<hpx::when_any_result<result_type>()> p(
+            [f = HPX_MOVE(f)]() -> hpx::when_any_result<result_type> {
                 return (*f)();
             });
 
+        auto result = p.get_future();
         p.apply();
-        return p.get_future();
+
+        return result;
     }
 
     template <typename Iterator,
-        typename Container = std::vector<
-            typename lcos::detail::future_iterator_traits<Iterator>::type>>
-    hpx::future<when_any_result<Container>> when_any(
+        typename Container =
+            std::vector<hpx::lcos::detail::future_iterator_traits_t<Iterator>>,
+        typename Enable =
+            std::enable_if_t<hpx::traits::is_iterator_v<Iterator>>>
+    hpx::future<hpx::when_any_result<Container>> when_any(
         Iterator begin, Iterator end)
     {
-        Container lazy_values_;
-
-        typename std::iterator_traits<Iterator>::difference_type difference =
-            std::distance(begin, end);
-        if (difference > 0)
-            traits::detail::reserve_if_reservable(
-                lazy_values_, static_cast<std::size_t>(difference));
+        Container values;
+        traits::detail::reserve_if_random_access_by_range(values, begin, end);
 
-        std::transform(begin, end, std::back_inserter(lazy_values_),
-            traits::acquire_future_disp());
-
-        return lcos::when_any(HPX_MOVE(lazy_values_));
+        std::move(begin, end, std::back_inserter(values));
+        return hpx::when_any(HPX_MOVE(values));
     }
 
-    inline hpx::future<when_any_result<hpx::tuple<>>>    //-V524
-    when_any()
+    inline auto when_any()
     {
-        typedef when_any_result<hpx::tuple<>> result_type;
-
-        return hpx::make_ready_future(result_type());
+        return hpx::make_ready_future(hpx::when_any_result<hpx::tuple<>>());
     }
 
     ///////////////////////////////////////////////////////////////////////////
     template <typename Iterator,
-        typename Container = std::vector<
-            typename lcos::detail::future_iterator_traits<Iterator>::type>>
-    hpx::future<when_any_result<Container>> when_any_n(
+        typename Container =
+            std::vector<hpx::lcos::detail::future_iterator_traits_t<Iterator>>,
+        typename Enable =
+            std::enable_if_t<hpx::traits::is_iterator_v<Iterator>>>
+    hpx::future<hpx::when_any_result<Container>> when_any_n(
         Iterator begin, std::size_t count)
     {
-        Container lazy_values_;
-        traits::detail::reserve_if_reservable(lazy_values_, count);
+        Container values;
+        traits::detail::reserve_if_reservable(values, count);
 
-        traits::acquire_future_disp func;
-        for (std::size_t i = 0; i != count; ++i)
-            lazy_values_.push_back(func(*begin++));
-
-        return lcos::when_any(lazy_values_);
+        while (count-- != 0)
+        {
+            values.push_back(HPX_MOVE(*begin++));
```

Comment:
Unfortunately we need this because clang-tidy is not able to reason about the macro definition correctly:
```suggestion
            // NOLINTNEXTLINE(bugprone-macro-repeated-side-effects)
            values.push_back(HPX_MOVE(*begin++));
```